// nuqs/dist/parsers.cjs
'use strict';

var chunkDAMYIQEW_cjs = require('./chunk-DAMYIQEW.cjs');
var chunkKNSPBEHI_cjs = require('./chunk-KNSPBEHI.cjs');

// src/index.parsers.ts
console.warn(
  "Please update your imports from `nuqs/parsers` to `nuqs/server`. Importing from `nuqs/parsers` is deprecated, and will be removed in v2.0.0."
);

Object.defineProperty(exports, "compareSearchParams", {
  enumerable: true,
  get: function () { return chunkDAMYIQEW_cjs.compareSearchParams; }
});
Object.defineProperty(exports, "createSearchParamsCache", {
  enumerable: true,
  get: function () { return chunkDAMYIQEW_cjs.createSearchParamsCache; }
});
Object.defineProperty(exports, "createParser", {
  enumerable: true,
  get: function () { return chunkKNSPBEHI_cjs.createParser; }
});
Object.defineProperty(exports, "createSerializer", {
  enumerable: true,
  get: function () { return chunkKNSPBEHI_cjs.createSerializer; }
});
Object.defineProperty(exports, "parseAsArrayOf", {
  enumerable: true,
  get: function () { return chunkKNSPBEHI_cjs.parseAsArrayOf; }
});
Object.defineProperty(exports, "parseAsBoolean", {
  enumerable: true,
  get: function () { return chunkKNSPBEHI_cjs.parseAsBoolean; }
});
Object.defineProperty(exports, "parseAsFloat", {
  enumerable: true,
  get: function () { return chunkKNSPBEHI_cjs.parseAsFloat; }
});
Object.defineProperty(exports, "parseAsHex", {
  enumerable: true,
  get: function () { return chunkKNSPBEHI_cjs.parseAsHex; }
});
Object.defineProperty(exports, "parseAsInteger", {
  enumerable: true,
  get: function () { return chunkKNSPBEHI_cjs.parseAsInteger; }
});
Object.defineProperty(exports, "parseAsIsoDateTime", {
  enumerable: true,
  get: function () { return chunkKNSPBEHI_cjs.parseAsIsoDateTime; }
});
Object.defineProperty(exports, "parseAsJson", {
  enumerable: true,
  get: function () { return chunkKNSPBEHI_cjs.parseAsJson; }
});
Object.defineProperty(exports, "parseAsNumberLiteral", {
  enumerable: true,
  get: function () { return chunkKNSPBEHI_cjs.parseAsNumberLiteral; }
});
Object.defineProperty(exports, "parseAsString", {
  enumerable: true,
  get: function () { return chunkKNSPBEHI_cjs.parseAsString; }
});
Object.defineProperty(exports, "parseAsStringEnum", {
  enumerable: true,
  get: function () { return chunkKNSPBEHI_cjs.parseAsStringEnum; }
});
Object.defineProperty(exports, "parseAsStringLiteral", {
  enumerable: true,
  get: function () { return chunkKNSPBEHI_cjs.parseAsStringLiteral; }
});
Object.defineProperty(exports, "parseAsTimestamp", {
  enumerable: true,
  get: function () { return chunkKNSPBEHI_cjs.parseAsTimestamp; }
});
